src/addr.rs

```diff
@@ -403,6 +403,108 @@ impl Step for VirtAddr {
     }
 }
 
+impl TryFrom<u64> for VirtAddr {
+    type Error = VirtAddrNotValid;
+
+    #[inline]
+    fn try_from(addr: u64) -> Result<Self, Self::Error> {
+        Self::try_new(addr)
+    }
+}
+
+// if `target_pointer_width > 64`, we would need a different error type
+#[cfg(any(
+    target_pointer_width = "16",
+    target_pointer_width = "32",
+    target_pointer_width = "64"
+))]
+impl TryFrom<usize> for VirtAddr {
+    type Error = VirtAddrNotValid;
+
+    #[inline]
+    fn try_from(addr: usize) -> Result<Self, Self::Error> {
+        Self::try_new(addr.try_into().unwrap())
+    }
+}
+
+// if `target_pointer_width > 64`, we would need a different error type
+// (and `as` would truncate)
+#[cfg(any(
+    target_pointer_width = "16",
+    target_pointer_width = "32",
+    target_pointer_width = "64"
+))]
+impl<T> TryFrom<*const T> for VirtAddr {
+    type Error = VirtAddrNotValid;
+
+    #[inline]
+    fn try_from(addr: *const T) -> Result<Self, Self::Error> {
+        Self::try_new(addr as u64)
+    }
+}
+
+// if `target_pointer_width > 64`, we would need a different error type
+// (and `as` would truncate)
+#[cfg(any(
+    target_pointer_width = "16",
+    target_pointer_width = "32",
+    target_pointer_width = "64"
+))]
+impl<T> TryFrom<*mut T> for VirtAddr {
+    type Error = VirtAddrNotValid;
+
+    #[inline]
+    fn try_from(addr: *mut T) -> Result<Self, Self::Error> {
+        Self::try_new(addr as u64)
+    }
+}
+
+impl From<u32> for VirtAddr {
+    #[inline]
+    fn from(addr: u32) -> Self {
+        Self::new(addr.into())
+    }
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+impl<T> From<&T> for VirtAddr {
+    #[inline]
+    fn from(addr: &T) -> Self {
+        Self::new(addr as *const _ as u64)
+    }
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+impl<T> From<&mut T> for VirtAddr {
+    #[inline]
+    fn from(addr: &mut T) -> Self {
+        Self::new(addr as *mut _ as u64)
+    }
+}
```

Comment:
These could technically fail if Intel's 5 level paging is active.

Reply:
This possibility was mentioned in issue #293, [here](https://github.com/rust-osdev/x86_64/issues/293#issuecomment-903098161) and [here](https://github.com/rust-osdev/x86_64/issues/293#issuecomment-903237326).

The last comment on the subject, by @josephlr was:
> Personally, I think the panic is fine (for now). Before we (potentially) support 5-level paging, we can just have a disclaimer that is like: "if you use 5-level paging, some methods might panic, but safety will not be violated".

No disclaimer has been added yet, but I can add one if that's the course we want to take.

Reply:
> > Personally, I think the panic is fine (for now). Before we (potentially) support 5-level paging, we can just have a disclaimer that is like: "if you use 5-level paging, some methods might panic, but safety will not be violated".

That seems like a good pragmatic option to me. Intel's LAM (Linear Address Masking) and AMD's UPA (Upper Address Ignore) can also cause problems because they relax the requirements on pointers to be canonical.
 
> No disclaimer has been added yet, but I can add one if that's the course we want to take.

That'd be great, thanks!